comp/metadata/runner/runner.go

```diff
@@ -0,0 +1,120 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package runner
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/comp/core/config"
+	"github.com/DataDog/datadog-agent/comp/core/log"
+	"go.uber.org/fx"
+)
+
+type MetadataProvider func(context.Context) time.Duration
+
+type runner struct {
+	log       log.Component
+	config    config.Component
+	providers []MetadataProvider
+
+	wg       sync.WaitGroup
+	stopChan chan struct{}
+}
+
+type dependencies struct {
+	fx.In
+
+	Log       log.Component
+	Config    config.Component
+	Providers []MetadataProvider `group:"metadata_provider"`
+}
+
+// Provider represents the callback from a metada provider. This is returned by 'NewProvider' helper.
+type Provider struct {
+	fx.Out
+
+	Callback MetadataProvider `group:"metadata_provider"`
+}
+
+// NewProvider registers a new metadata provider by adding a callback to the runner.
+func NewProvider(callback MetadataProvider) Provider {
+	return Provider{
+		Callback: callback,
+	}
+}
+
+// createRunner instantiates a runner object
+func createRunner(deps dependencies) *runner {
+	return &runner{
+		log:       deps.Log,
+		config:    deps.Config,
+		providers: deps.Providers,
+		stopChan:  make(chan struct{}),
+	}
+}
+
+func newRunner(lc fx.Lifecycle, deps dependencies) Component {
+	r := createRunner(deps)
+
+	// We rely on FX to start and stop the metadata runner
+	lc.Append(fx.StartStopHook(r.start, r.stop))
+	return r
+}
+
+// handleProvider runs a provider at regular interval until the runner is stopped
+func (r *runner) handleProvider(p MetadataProvider) {
+	r.log.Debugf("Starting runner for MetadataProvider %v", p)
+
+	intervalChan := make(chan time.Duration)
+	var interval time.Duration
+
+	ctx, cancel := context.WithCancel(context.Background())
+	for {
+		go func(intervalChan chan time.Duration) {
+			intervalChan <- p(ctx)
+		}(intervalChan)
+
+		select {
+		case interval = <-intervalChan:
+			cancel()
```

Comment:
Why calling `cancel()` on the context when the provider returns a value? Isn't `handleProvider` supposed to be called several times?